### PR TITLE
Add Falcon3 support and Fix issue #10875

### PR DESCRIPTION
**Pull request by mokeddembillel**
This PR adds Falcon3 support and fixes issue #10875  caused by previous PR #10864 (see #10864 for details) 

Details of fixing issue #10875:

The issue is that when using meta-llama/Llama-3.1-8B-Instruct the <|begin_of_text|> token is added to every special token when doing `token = tokenizer.decode(tokenizer.encode(token))`
 
the screenshot shows before and after `token = tokenizer.decode(tokenizer.encode(token))`
image

I'm fixing this by adding `add_special_tokens=False` to `tokenizer.encode()`. Here is the the result after the fix
image

to be extra safe, we will use `token = tokenizer.decode(tokenizer.encode(token))` only if `len(token) == 1` so that still fix this issue when `\n` is econded as `Ċ`

Generation before the fix:
```
Prompt: Once upon a time in a land far away,
there was a kingdom ruled by a wise and just king. The kingdom was known for its beauty and prosperity, and the people lived in peace and harmony.ĊĊOne day, a terrible drought struck the land, and the crops began to wither and die. The king, worried about the well-being of his people, called upon his wise council to find a solution. The council, after much deliberation, decided to send a group of brave knights to search for a magical spring that was said to have the power to bring rain to the kingdom.
```
Generation after the fix:
```
Prompt: Once upon a time in a land far away,
there was a kingdom ruled by a wise and just king. The kingdom was known for its beauty and prosperity, and the people lived in peace and harmony.

One day, a terrible drought struck the land, and the crops began to wither and die. The king, worried about the well-being of his people, called upon his wise council to find a solution. The council, after much deliberation, decided to send a group of brave knights to search for a magical spring that was said to have the power to bring rain to the kingdom.
```

@ggerganov @compilade @slaren

### CODE REVIEW

File: convert_hf_to_gguf.py
Comment (younesbelkada) [inline convert_hf_to_gguf.py:530]:
```suggestion
                    # To avoid unexpected issues - we make sure to encode single-char tokens
                    if len(token) == 1:
```

**Review by younesbelkada — APPROVED** [review]

File: convert_hf_to_gguf.py
Comment (ggerganov) [reply convert_hf_to_gguf.py:531]: I'm looking at the Falcon tokenizer and I don't see any added tokens that have `\n` or `\t`: https://huggingface.co/tiiuae/Falcon3-7B-Instruct/raw/main/tokenizer.json

For which tokens does this change make a difference?

Maybe also add some logs to know when this path is being triggered so we can spot any potential problems with other models.

Comment (younesbelkada) [reply convert_hf_to_gguf.py:531]: Chiming in here ! The added token is 
```
    {
      "id": 12,
      "content": "Ċ",
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": false,
      "special": false
    }
``` (`\t` is the id 13)
the only way to convert it properly to `\n` is to encode / decode using the tokenizer

Comment (mokeddembillel) [reply convert_hf_to_gguf.py:531]: I just added a log message inside the if statement.

File: convert_hf_to_gguf.py
Comment (ggerganov) [inline convert_hf_to_gguf.py:532]: I was thinking about comparing the token before and after the encoding and print the log only if there is a difference.

Comment (mokeddembillel) [reply convert_hf_to_gguf.py]: that's a good idea. Done!

**Review by ggerganov — APPROVED** [review]
Seems OK to me, but I am not sure about the full implications of this change for all other models. Want to wait for some feedback from the community.

The alternative is to find a way to apply this logic only inside the `class FalconModel`.

**Comment by mokeddembillel** [comment]
Actually there's no FalconModel class and our model type is llama so we can't use that to check. The only solution I see is that we wait for some feedback from the community and if anything related to this, I will be happy to address it and fix it quickly.

**Comment by compilade** [comment]
> but I am not sure about the full implications of this change for all other models.

This can be tested by converting all tokenizers fetched by `convert_hf_to_gguf_update.py` and comparing the hashes when converted before and after this change (which I can't do right now, but will when I can).

I think what would solve variations of this problem for other models in the future (for another PR) would be to either normalize all added tokens which are marked `"normalized": false`, since the added tokens are internally assumed to be pre-normalized (this is the same problem which #8228 attempted to solve, but apparently the fix wasn't general enough), *or* non-normalized added tokens could be internally handled by adding some token attribute for it. Though this would depend on proper support for token attributes stored in GGUF files, which isn't yet complete (even though per-token attributes were added in <https://github.com/ggerganov/llama.cpp/pull/7685>, they aren't stored in GGUF models, and `LLAMA_TOKEN_ATTR_NORMALIZED` isn't really handled).

**Comment by younesbelkada** [comment]
Thanks @compilade !
There might be an easier solution, I am about to manually modify the normalized characters (since the problem is only for \n and \t that have been explciitly added as special tokens) in the tokenizer file and push the normalized tokenizer on all repos - then we can convert this PR to simply adding the falcon3 pre-tokenizer - what do you think?

**Comment by compilade** [comment]
> There might be an easier solution, I am about to manually modify the normalized characters (since the problem is only for \n and \t that have been explciitly added as special tokens) in the tokenizer file and push the normalized tokenizer on all repos - then we can convert this PR to simply adding the falcon3 pre-tokenizer - what do you think?

@younesbelkada 

That could also work, as long as it's done correctly. The added tokens are in both `tokenizers.json` and `tokenizer_config.json`. If you do this, make sure that it doesn't have unintended consequences.

This is otherwise a nice edge case I think the convert scripts *should* have handled correctly, so part of me wants to keep the tokenizers the same.

**Comment by younesbelkada** [comment]
Perfect thanks, will test that out and update here !

**Comment by younesbelkada** [comment]
@compilade I just did some tests and I think we can't go with the solution I suggested above - mainly due to backward compatibility reasons - Before the manual changes
```
(Pdb) tok.encode("ĉ")
[13]
```
After the fix I suggested: 
```
(Pdb) fixed_tokenizer.encode("ĉ")
[2150, 2237]
```
--> For the same token we now get different encodings - As all falcon3 series models have been trained with that tokenizer, even if the probably that this token appears in a text, I am afraid it's a way too risky breaking change to introduce.. 
I also tried to set `normalize: true` to these tokens and converted the model with this PR and still getting `Ċ` printed all over the place for line breaks.

Perhaps we can test if existing tokenizers are not affected by this PR, what do you think?

**Comment by xhatz** [comment]
Hey, I'm still getting this error:

`error loading model vocabulary: unknown pre-tokenizer type: 'falcon3'`

Is it normal? I've read on a previous llamacpp version that is was updated to handle it. Thanks!

**Comment by mokeddembillel** [comment]
@xhatz we need to wait until this pull request is merged. 

@compilade is there any way we can help to merge this pull request faster? Thanks

**Comment by xhatz** [comment]
Oh alright, thanks! I had read it was already merged in b4341 that's why I was confused haha.

**Comment by ggerganov** [comment]
> This can be tested by converting all tokenizers fetched by convert_hf_to_gguf_update.py and comparing the hashes when converted before and after this change (which I can't do right now, but will when I can).

Which hashes do you have in mind? The hashes produced by `convert_hf_to_gguf_update.py` are generated by the original Python tokenizer - it is not affected by this change, so they would be the same.

I reconverted all tokenizers that I have access to:

```bash
python3.11 convert_hf_to_gguf.py models/tokenizers/llama-spm/ --outfile models/ggml-vocab-llama-spm.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/llama-bpe/ --outfile models/ggml-vocab-llama-bpe.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/phi-3/ --outfile models/ggml-vocab-phi-3.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/deepseek-llm/ --outfile models/ggml-vocab-deepseek-llm.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/deepseek-coder/ --outfile models/ggml-vocab-deepseek-coder.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/falcon/ --outfile models/ggml-vocab-falcon.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/bert-bge/ --outfile models/ggml-vocab-bert-bge.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/falcon3/ --outfile models/ggml-vocab-falcon3.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/bert-bge-large/ --outfile models/ggml-vocab-bert-bge-large.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/mpt/ --outfile models/ggml-vocab-mpt.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/starcoder/ --outfile models/ggml-vocab-starcoder.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/gpt-2/ --outfile models/ggml-vocab-gpt-2.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/stablelm2/ --outfile models/ggml-vocab-stablelm2.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/refact/ --outfile models/ggml-vocab-refact.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/command-r/ --outfile models/ggml-vocab-command-r.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/qwen2/ --outfile models/ggml-vocab-qwen2.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/olmo/ --outfile models/ggml-vocab-olmo.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/dbrx/ --outfile models/ggml-vocab-dbrx.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/jina-v1-en/ --outfile models/ggml-vocab-jina-v1-en.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/jina-v2-en/ --outfile models/ggml-vocab-jina-v2-en.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/jina-v2-es/ --outfile models/ggml-vocab-jina-v2-es.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/jina-v2-de/ --outfile models/ggml-vocab-jina-v2-de.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/smaug-bpe/ --outfile models/ggml-vocab-smaug-bpe.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/poro-chat/ --outfile models/ggml-vocab-poro-chat.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/jina-v2-code/ --outfile models/ggml-vocab-jina-v2-code.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/viking/ --outfile models/ggml-vocab-viking.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/gemma/ --outfile models/ggml-vocab-gemma.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/gemma-2/ --outfile models/ggml-vocab-gemma-2.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/t5/ --outfile models/ggml-vocab-t5.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/codeshell/ --outfile models/ggml-vocab-codeshell.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/tekken/ --outfile models/ggml-vocab-tekken.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/smollm/ --outfile models/ggml-vocab-smollm.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/bloom/ --outfile models/ggml-vocab-bloom.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/gpt3-finnish/ --outfile models/ggml-vocab-gpt3-finnish.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/phi-2/ --outfile models/ggml-vocab-phi-2.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/minerva-7b/ --outfile models/ggml-vocab-minerva-7b.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/roberta-bpe/ --outfile models/ggml-vocab-roberta-bpe.gguf --vocab-only
python3.11 convert_hf_to_gguf.py models/tokenizers/gigachat/ --outfile models/ggml-vocab-gigachat.gguf --vocab-only
```

The new log message was triggered for the following tokenizers:

```
INFO:hf-to-gguf:Loading model: deepseek-llm
INFO:hf-to-gguf:'ø' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ö' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ú' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ÿ' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'õ' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'÷' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'û' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ý' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'À' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ù' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'Á' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'þ' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ü' is encoded and decoded back to '�' using AutoTokenizer

INFO:hf-to-gguf:Loading model: deepseek-coder
INFO:hf-to-gguf:'õ' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'÷' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'Á' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ý' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'À' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ÿ' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ø' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ú' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'þ' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ü' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ù' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'ö' is encoded and decoded back to '�' using AutoTokenizer
INFO:hf-to-gguf:'û' is encoded and decoded back to '�' using AutoTokenizer

INFO:hf-to-gguf:Loading model: falcon3
INFO:hf-to-gguf:'Ċ' is encoded and decoded back to '\n' using AutoTokenizer
INFO:hf-to-gguf:'ĉ' is encoded and decoded back to '\t' using AutoTokenizer
```

So it seems Deepseek models will be affected and I cannot tell if the change is good or bad. Any ideas?

Also, there might be more models affected for which I haven't requested access yet: `jais`, `exaone`, `chameleon`.

**Comment by mokeddembillel** [comment]
I'm looking into it. Thanks!

**Comment by compilade** [comment]
> Which hashes do you have in mind? The hashes produced by `convert_hf_to_gguf_update.py` are generated by the original Python tokenizer - it is not affected by this change, so they would be the same.

I meant the hashes of the GGUF vocab-only models made when reconverting the tokenizers with `convert_hf_to_gguf.py`.

<details><summary>I've done this hash comparison test for the below suggested change.</summary>

```console
$ sha256sum --check tokenizers.SHA256SUMS 
models/ggml-vocab-aquila.gguf: OK
models/ggml-vocab-baichuan.gguf: OK
models/ggml-vocab-bert-bge.gguf: OK
models/ggml-vocab-bert-bge-large.gguf: OK
models/ggml-vocab-bloom.gguf: OK
models/ggml-vocab-codeshell.gguf: OK
models/ggml-vocab-command-r.gguf: OK
models/ggml-vocab-dbrx.gguf: OK
models/ggml-vocab-deepseek-coder.gguf: OK
models/ggml-vocab-deepseek-llm.gguf: OK
models/ggml-vocab-falcon3.gguf: FAILED
models/ggml-vocab-falcon.gguf: OK
models/ggml-vocab-gemma-2.gguf: OK
models/ggml-vocab-gemma.gguf: OK
models/ggml-vocab-gigachat.gguf: OK
models/ggml-vocab-gpt-2.gguf: OK
models/ggml-vocab-gpt3-finnish.gguf: OK
models/ggml-vocab-gpt-neox.gguf: OK
models/ggml-vocab-jais.gguf: OK
models/ggml-vocab-jina-v1-en.gguf: OK
models/ggml-vocab-jina-v2-code.gguf: OK
models/ggml-vocab-jina-v2-de.gguf: OK
models/ggml-vocab-jina-v2-en.gguf: OK
models/ggml-vocab-jina-v2-es.gguf: OK
models/ggml-vocab-llama-bpe.gguf: OK
models/ggml-vocab-llama-spm.gguf: OK
models/ggml-vocab-minerva-7b.gguf: OK
models/ggml-vocab-mpt.gguf: OK
models/ggml-vocab-olmo.gguf: OK
models/ggml-vocab-phi-2.gguf: OK
models/ggml-vocab-phi-3.gguf: OK
models/ggml-vocab-poro-chat.gguf: OK
models/ggml-vocab-qwen2.gguf: OK
models/ggml-vocab-refact.gguf: OK
models/ggml-vocab-roberta-bpe.gguf: OK
models/ggml-vocab-smaug-bpe.gguf: OK
models/ggml-vocab-smollm.gguf: OK
models/ggml-vocab-stablelm2.gguf: OK
models/ggml-vocab-starcoder.gguf: OK
models/ggml-vocab-t5.gguf: OK
models/ggml-vocab-tekken.gguf: OK
models/ggml-vocab-viking.gguf: OK
sha256sum: WARNING: 1 computed checksum did NOT match
```

And only `falcon3` doesn't match, which is expected.

---

</details>

> I think what would solve variations of this problem for other models in the future (for another PR) would be to either normalize all added tokens which are marked `"normalized": false`,

I've implemented this suggestion, and I think it works correctly.

```diff
diff --git a/convert_hf_to_gguf.py b/convert_hf_to_gguf.py
index 06e3016c..894c3e4d 100755
--- a/convert_hf_to_gguf.py
+++ b/convert_hf_to_gguf.py
@@ -525,10 +525,9 @@ class Model:
             else:
                 token: str = reverse_vocab[i]
                 if token in added_vocab:
-                    # We need to manually encode and decode the added tokens in case special characters
-                    # used for `\n` / `\t` have been manually added in the added tokens
-                    # To avoid unexpected issues - we make sure to encode single-char tokens
-                    if len(token) == 1:
+                    # The tokenizer in llama.cpp assumes the CONTROL and USER_DEFINED tokens are pre-normalized.
+                    # To avoid unexpected issues - we make sure to normalize non-normalized tokens
+                    if not tokenizer.added_tokens_decoder[i].normalized:
                         previous_token = token
                         token = tokenizer.decode(tokenizer.encode(token, add_special_tokens=False))
                         if previous_token != token:
@@ -537,6 +536,8 @@ class Model:
                     if tokenizer.added_tokens_decoder[i].special or self.does_token_look_special(token):
                         toktypes.append(gguf.TokenType.CONTROL)
                     else:
+                        # NOTE: this was added for Gemma.
+                        # Encoding and decoding the tokens above isn't sufficient for this case.
                         token = token.replace(b"\xe2\x96\x81".decode("utf-8"), " ")  # pre-normalize user-defined spaces
                         toktypes.append(gguf.TokenType.USER_DEFINED)
                 else:
```

The deepseek models are not affected anymore, since they properly mark their tokens as normalized.

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Nice. @mokeddembillel After applying the patch by @compilade we can merge.

**Comment by mokeddembillel** [comment]
Thanks a lot for the help @compilade and @ggerganov. Really appreciated🙏🏼

**Comment by ggerganov** [comment]
No problem - do you want me to apply the patch, or will you do it?

**Comment by mokeddembillel** [comment]
Oh yes for sure. I will do it as soon as I get home. Thanks 🙏🏼

**Review by compilade — APPROVED** [review]

File: src/llama.cpp
Comment (ggerganov) [inline src/llama.cpp:6482]: Let's simplify this by moving the check to the `LLAMA3` branch above:

```c++
                    tokenizer_pre == "llama3"    ||
                    tokenizer_pre == "llama-v3"  ||
                    tokenizer_pre == "llama-bpe" ||
                    tokenizer_pre == "falcon3") {
```

Comment (mokeddembillel) [reply src/llama.cpp]: Done.

**Pull request merged**
